ml-prims/src/common/grid_sync.h

```diff
@@ -64,6 +64,41 @@ enum SyncType {
  * CUDA_CHECK(cudaFree(workspace));
  * @endcode
  *
+ * @note Calling `GridSync::sync` method consecutively on the same object inside
+ * the same kernel can lead to deadlocks. Such a usage is discouraged. Also,
+ * trying to use different 'SyncType' in the same kernel call can also lead to
+ * deadlocks and thus is also discouraged.
+ *
+ * @note In case you need more than one sync in the same kernel call, better to
+ * create 2 different GridSync objects with different workspaces and ping-pong
+ * between the two. Example follows:
+ *
+ * @code{.cu}
+ * __global__ void kernelMultiple(void* workspace1, void* workspace2,
+ *                                SyncType type, ...) {
+ *   GridSync gs1(workspace1, type);
+ *   GridSync gs2(workspace2, type);
+ *   ////// Part1 //////
+ *   // do pre-sync work here
+ *   // ...
+ *   gs1.sync();
+ *   // do post-sync work here
+ *   // ...
+ *   ////// Part2 //////
+ *   // do pre-sync work here
+ *   // ...
+ *   gs2.sync();
+ *   // do post-sync work here
+ *   // ...
+ *   ////// Part3 //////
+ *   // do pre-sync work here
+ *   // ...
+ *   gs1.sync();
+ *   // do post-sync work here
+ *   // ...
+ * }
+ * @endcode
```

Comment:
The two object approach would require additional resources which might be an issue for persistent kernels. If we instead have a single object with two workspaces and swap them at the end of `waitForOthers`, we have the same behavior as the two object approach and avoid data races. This way, use of the grid sync is also more intuitive at the call site.

Reply:
Can change constructor to `DI GridSync(void* workspace, SyncType _type, bool multipleSync = false)`  so that we can only have one workspace for the common use case of a single sync per kernel.

Reply:
Done.